### PR TITLE
Horizontal scroll support

### PR DESCRIPTION
**Pull request by can3p**
This pull request adds support for horizontal mouse scroll in serenity.  I'm not sure if that's the best or the most correct implementation, but it works.

Information on horizontal scroll for ps/2 mice: https://wiki.osdev.org/Mouse_Input#Formats_of_Optional_4th_Packet_Byte

In order to implement the support I had to hack qemu too since it also doesn't support horizontal scroll.

The branch with changes: https://gitlab.com/can3p/qemu/-/commits/mouse-horizontal-scroll

The way I ended up implementing it looks hackish to me (not sure where `0x3f` from `b = (dw1 & 0x3f) | 0x40;
` comes from. Anyway, it seems to be working for ubuntu vm which I used as a reference.

This pr works in bth gtk and sdl display modes, with and without vmware hack.

I haven't pushed submitted the patch to qemu project yet, planning to do that after testing the cocoa related code in the patch.

Limitation of this pr and qemu patch: scroll is only possible in one direction (horizontal or vertical) whereas ubuntu system which I take as a reference allows to scroll the window in combination of both if necessary

### CODE REVIEW

File: Kernel/Devices/HID/PS2MouseDevice.cpp [hunk 1]
```diff
@@ -97,6 +98,14 @@ MousePacket PS2MouseDevice::parse_data_packet(const RawPacket& raw_packet)
         // -1 in 4 bits
         if (z == 15)
             z = -1;
+
+        if ((raw_packet.bytes[3] & 0xc0) == 0x40) {
+            // FIXME: obviously I've messed up signs somehwere
```
Comment (kleinesfilmroellchen) [inline Kernel/Devices/HID/PS2MouseDevice.cpp:103]: Typo; also, please rephrase this to not refer to "you" specifically. Once the code is in, it's everyone's code and everyone may fix the issue.

File: Kernel/Devices/VMWareBackdoor.cpp [hunk 2]
```diff
@@ -211,16 +211,28 @@ MousePacket VMWareBackdoor::receive_mouse_packet()
     int x = command.bx;
     int y = command.cx;
     int z = static_cast<i8>(command.dx); // signed 8 bit value only!
+    int w = 0;
+
+    // horizontal scroll is reported as +-2 by qemu
+    // I've probably messed up the sign there too, hence reversed values
```
Comment (kleinesfilmroellchen) [inline Kernel/Devices/VMWareBackdoor.cpp:217]: (see above)

File: Userland/Services/WindowServer/EventLoop.cpp [hunk 3]
```diff
@@ -64,7 +64,7 @@ void EventLoop::drain_mouse()
     bool state_is_sent = false;
     for (size_t i = 0; i < npackets; ++i) {
         auto& packet = packets[i];
-        dbgln_if(WSMESSAGELOOP_DEBUG, "EventLoop: Mouse X {}, Y {}, Z {}, relative={}", packet.x, packet.y, packet.z, packet.is_relative);
+        dbgln_if(WSMESSAGELOOP_DEBUG, "EventLoop: Mouse X {}, Y {}, Z {}, W {} relative={}", packet.x, packet.y, packet.z, packet.w, packet.is_relative);
```
Comment (kleinesfilmroellchen) [inline Userland/Services/WindowServer/EventLoop.cpp:67]: Keeping a comma after each reported value would be nice

File: Userland/Libraries/LibGUI/WindowServerConnection.cpp [hunk 4]
```diff
@@ -227,38 +227,53 @@ static MouseButton to_mouse_button(u32 button)
...
+    switch (direction) {
+    case 0:
+        return WheelDirection::WheelNone;
+    case 1:
+        return WheelDirection::WheelVertical;
+    case 2:
+        return WheelDirection::WheelHorizontal;
+    default:
+        VERIFY_NOT_REACHED();
+        break;
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/WindowServerConnection.cpp:241]: This is unnecessary after a VERIFY.

File: Userland/Libraries/LibGUI/Event.h [hunk 5]
```diff
@@ -318,6 +318,12 @@ enum MouseButton : u8 {
     Forward = 16,
 };
 
+enum WheelDirection : u8 {
+    WheelNone = 0,
+    WheelVertical = 1,
+    WheelHorizontal = 2,
+};
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/Event.h:325]: Why did you choose a plain enum here? After all, in the window server there's an enum class...

File: Userland/Services/WindowServer/Screen.cpp [hunk 6]
```diff
@@ -462,7 +462,10 @@ void ScreenInput::on_receive_mouse_data(const MousePacket& packet)
     }
 
     if (packet.z) {
-        auto message = make<MouseEvent>(Event::MouseWheel, m_cursor_location, buttons, MouseButton::None, m_modifiers, packet.z * m_scroll_step_size);
+        auto message = make<MouseEvent>(Event::MouseWheel, m_cursor_location, buttons, MouseButton::None, m_modifiers, packet.z * m_scroll_step_size, WheelDirection::Vertical);
+        Core::EventLoop::current().post_event(WindowManager::the(), move(message));
+    } else if (packet.w) {
+        auto message = make<MouseEvent>(Event::MouseWheel, m_cursor_location, buttons, MouseButton::None, m_modifiers, packet.w * m_scroll_step_size, WheelDirection::Horizontal);
         Core::EventLoop::current().post_event(WindowManager::the(), move(message));
```
Comment (kleinesfilmroellchen) [inline Userland/Services/WindowServer/Screen.cpp:469]: As this is common code, it may be cleaner to move it outside of the conditionals (even if you have to work around some stuff with the variable).

File: Userland/Libraries/LibGUI/Event.h (hunk 5 above)
Comment (can3p) [reply Userland/Libraries/LibGUI/Event.h:325]: To be honest I've just mimicked mouse button enum that's defined in the same fashion above this one

**Comment by can3p** [comment]
@kleinesfilmroellchen Thanks for the comments! I've fixed everything except enum. Do you mean reuse the same enum in both window server and libgui code? I didn't do it just to follow the way two MouseButton enums were defined in two places

**Comment by kleinesfilmroellchen** [comment]
Although it may be possible somehow, reusing enums weren't my intention. I would just use the same style of enum in both cases so that the code reads the same in both locations where wheel directions are handled.

**Review by kleinesfilmroellchen — APPROVED** [review]
As long as you address the final comment, looks good to me

**Comment by can3p** [comment]
@kleinesfilmroellchen I think now I got what I needed to fix, didn't know about enum vs enum class in c++. Thank you for the comment! I've pushed a change to make the definition of both enums completely identical

I'll submit the patch to qemu soon. I think it's fair to wait before it's accepted, since it's not possible to actually test this pr otherwise

**Review by kleinesfilmroellchen — APPROVED** [review]
Looks good to me. Let's wait on the upstream QEMU patch; you'll have to remember to rebase regularly in the meantime to avoid the stale bot.

**Comment by can3p** [comment]
Qemu support is in master! https://gitlab.com/qemu-project/qemu/-/compare/f0602b709985b28bc8ce572d6c517d9c26669735...17f6315ef883a142b6a41a491b63a6554e784a5c?from_project_id=11167699

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by awesomekling** [comment]
Hmm, would it make more sense for the wheel events to have both vertical and horizontal deltas?

**Comment by can3p** [comment]
@awesomekling That's one of the things I wasn't sure about, mostly because I was only able to get wheel events in one direction only, but indeed having both deltas would be more general.

I'll have a look tonight if I can rework it quickly

**Comment by can3p** [comment]
@awesomekling I've updated the pr to pass wheel values for both axis. Quite an update but it allowed to catch a few places in lib web that were expecting horizontal scroll but it was hardcoded as zero. I've triple checked the change and tested it in browser and text editor, but still not sure if I was able to test all the cases. Thanks for the review!

**Comment by awesomekling** [comment]
Sweet! :star:
